<a href="https://colab.research.google.com/github/luislauriano/Alura_QuarentenaDados/blob/master/Aula1_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Manipulação de Dados**

In [0]:
import pandas as pd

# **Leitura de Arquivos**

## Arquivos CSV e TXT

pd.read_csv(**string** , sep = **string** , skiprows = **int**, encoding = **string específica**)

1. nome do arquivo
2. string que separa os dados no arquivo
3. quantidade de linhas a serem puladas
4. como o arquivo está codificado: utf-8 ou latin-1

In [0]:
df_materiais = pd.read_csv('materiais.csv', sep = ';', skiprows = 0, encoding = 'utf-8')

In [6]:
df_materiais

Material     Texto breve material
0      4000010         ALCACHOFRA 300MG
1      4000013  ANTI-SEPTICO ANESTESICO
2      4000014           ASSEPTCARE SOL
3      4000015           ASSEPTCARE SOL
4      4000016        BENATUX FRAMBOESA
...        ...                      ...
4577   6001409             SABONETE LIQ
4578   6001410             SABONETE LIQ
4579   6001411             TÔNICO NIVEA
4580   6001412             TÔNICO NIVEA
4581   6001415         CHOCOLATE KINDER

[4582 rows x 2 columns]

In [7]:
df_materiais['Material'].loc[0]

4000010

In [0]:
df_pedidos = pd.read_csv('pedidos.txt' , sep= '|', skiprows = 2, encoding = 'latin-1')

In [9]:
df_pedidos 

Cod. fornecedor   Material  Qtd.pedido      Texto breve  Qtd.estoque
0                3593  4003232.0         120   FLUCANIL 150MG           12
1                3593  4005091.0          60  BETRICORT CREME            6
2                3593  4003225.0          51   COLIRIO GEOLAB            5
3                3593  4003251.0          60    MIZONOL CREME            6
4                3593  4001716.0          20  GINO-COLON TUBO            2
...               ...        ...         ...              ...          ...
1064             4239  4000333.0           1     CRESTOR 10MG            0
1065             4239  4003698.0           1  PANTOGAR C/90CP            0
1066             4239  4000359.0           1     XARELTO 15MG            0
1067             4239  4005114.0           1        ARADOIS H            0
1068             4239  4003731.0           1     DIOVAN 160MG            0

[1069 rows x 5 columns]

## Arquivos XLSX

pd.read_excel( **string**, sheet_name = **string** ou **int**, skiprows = **int**)

1. nome do arquivo
2. nome da aba do excel ou o indice da aba do excel
3. número de linhas a serem puladas

In [0]:
df_vendas1 = pd.read_excel('vendas_1.xlsx', sheet_name = 'vendas', skiprows = 0)

In [11]:
df_vendas1

Data  Material  Qtd.faturd
0     25-04-2020   6000110           1
1     25-04-2020   6000129           1
2     25-04-2020   6000163           1
3     25-04-2020   6000172           1
4     25-04-2020   6000892           1
...          ...       ...         ...
8652  27-04-2020   4005739           1
8653  27-04-2020   4005739           1
8654  27-04-2020   4005903           1
8655  27-04-2020   4005911           1
8656  27-04-2020   4005927           1

[8657 rows x 3 columns]

Também poderiamos colocar o sheet_name como indice, iniciando a ordem com 0

In [0]:
df_vendas2 = pd.read_excel('vendas_2.xlsx', sheet_name = 0 , skiprows = 0)

# **Tratamento de Dados**

## Renomeando colunas

**Posso renomear uma ou mais colunas com um dicionario**

dataframe.rename(columns=**dict**)

1. dict = {'coluna_antiga':'coluna_nova'}

In [0]:
df_materiais = df_materiais.rename(columns={'Material':'material', 'Texto breve material':'texto_breve'})

**Ou renomear todas as colunas como uma lista**

dataframe.columns = **lista**

1. lista = ['coluna_nova']

In [0]:
df_pedidos.columns = ['cod_fornecedor','material', 'qtd_pedido', 'texto_breve', 'qtd_estoque']

In [0]:
df_vendas1.columns = ['data', 'material', 'qtd_vendas']

In [0]:
df_vendas2.columns = ['data', 'material', 'qtd_vendas']

## Deletando colunas

dataframe.drop( **lista**, axis = **string específica** ou int)

1. lista de nomes de colunas ou index das linhas a serem deletadas

2. axis = 'columns' p/ colunas ou 'index' p/ linhas
2. Ou 1 (colunas) ou 0 (linhas)

In [0]:
df_materiais = df_materiais.drop( ['texto_breve'], axis = 'columns')

Agora vamos ver como se faz para escolher apenas as colunas que desejamos utilizar e outras colunas serão esquecidas, mas não deletadas

Esse formato também permite escolher a ordem das colunas

dataframe[**lista**].copy()

1. lista = ['colunas a manter']

In [0]:
df_pedidos = df_pedidos[['material', 'qtd_pedido', 'qtd_estoque']].copy()

In [0]:
df_vendas1 = df_vendas1[['material','qtd_vendas']].copy()

In [0]:
df_vendas2 = df_vendas2.drop( ['data'], axis = 'columns')

## Verificando tipos de colunas

dataframe.dtypes (para verificar tipos de todas as colunas)

dataframe['coluna'].dtype

In [21]:
df_materiais.dtypes

material    int64
dtype: object

In [22]:
df_pedidos.dtypes

material       float64
qtd_pedido       int64
qtd_estoque      int64
dtype: object

In [23]:
df_vendas1.dtypes

material      int64
qtd_vendas    int64
dtype: object

In [24]:
df_vendas2.dtypes

material      int64
qtd_vendas    int64
dtype: object

Pra visualizar o tipo de uma coluna especifica:


In [25]:
df_vendas2['material'].dtypes

dtype('int64')

## Alterando tipos de colunas

para apenas uma coluna:

dataframe['coluna'].astype(palavra reservada)

1. palavra reservada = str ou int ou float ou bool

In [0]:
df_pedidos['material'] = df_pedidos['material'].astype(int)

Agora vamos escolher as nossas colunas e já transformar os seus tipos

In [0]:
df_pedidos[['qtd_pedido','qtd_estoque']] = df_pedidos[['qtd_pedido','qtd_estoque']].astype(int)

## Concatenação de tabelas (Juntar tabelas)

pd.concat(**lista**)

1. lista = [lista de dataframes]

In [28]:
df_vendas1.shape

(8657, 2)

In [29]:
df_vendas2.shape

(8144, 2)

In [0]:
df_vendas = pd.concat([df_vendas1,df_vendas2])

In [31]:
df_vendas

material  qtd_vendas
0      6000110           1
1      6000129           1
2      6000163           1
3      6000172           1
4      6000892           1
...        ...         ...
8139   6000768           1
8140   6000839           1
8141   6000843           1
8142   6000969           1
8143   6000400           1

[16801 rows x 2 columns]

Temos 16801 linhas, porem nosso index esta ate o 8143

Os index precisam ser zerados e voltarem para a ordem padrao 


### Resetando o index

dataframe.reset_index(drop=**bool**)

1. True ou False

In [0]:
df_vendas = df_vendas.reset_index(drop=True)

In [33]:
df_vendas

material  qtd_vendas
0       6000110           1
1       6000129           1
2       6000163           1
3       6000172           1
4       6000892           1
...         ...         ...
16796   6000768           1
16797   6000839           1
16798   6000843           1
16799   6000969           1
16800   6000400           1

[16801 rows x 2 columns]

## Agrupando dados

dataframe.groupby(**lista**).agg(**dict**).reset_index()

1. lista = [colunas a serem agrupadas]
2. dict = {'coluna_x':'funcao'}

Agora para aqueles materiais(linhas) iguais  que estao se repetindo após a concatenação, vamos somar os seus valores e tranformar em unico registro (linha). 

Vamos usar o **groupby** para agrupar e **sum** para somar

In [0]:
df_vendas = df_vendas.groupby(['material']).agg({'qtd_vendas':'sum'}).reset_index(drop=False)

In [0]:
df_pedidos = df_pedidos.groupby(['material']).agg({'qtd_pedido':'sum','qtd_estoque':'sum'}).reset_index(drop=False)

In [36]:
df_vendas.head()

material  qtd_vendas
0   4000013           6
1   4000014           1
2   4000015           8
3   4000016           3
4   4000017           6

In [37]:
df_pedidos.head()

material  qtd_pedido  qtd_estoque
0   4000010         250           25
1   4000014          60            6
2   4000019           7            0
3   4000020           9            0
4   4000047          50            5